### ISSUE TITLE
what is the correct way to use with postcss-purgecss ?

### ISSUE THREAD

**Issue by vesper8** [bootstrap-vue/bootstrap-vue#5082]
This is my current `postcss.config.js`

```
module.exports = {
  plugins: [
    require('tailwindcss')('./tailwind.config.js'),
    require('autoprefixer')({}),
    process.env.NODE_ENV === 'production' ? require('@fullhuman/postcss-purgecss')({
      content: [
        './public/**/*.html',
        './src/**/*.vue',
        // custom
        './node_modules/bootstrap-vue/src/components/**/*.js',
      ],
      defaultExtractor: (content) => {
        const contentWithoutStyleBlocks = content.replace(/<style[^]+?<\/style>/gi, '');
        return contentWithoutStyleBlocks.match(/[A-Za-z0-9-_/:]*[A-Za-z0-9-_/]+/g) || [];
      },
      whitelistPatterns: [
        /-(leave|enter|appear)(|-(to|from|active))$/,
        /^(?!cursor-move).+-move$/,
        /^router-link(|-exact)-active$/,
        // custom
        /^pac/, // google-maps autocomplete
        /^multiselect/,
        /^shepherd/,
        /^swal2/,
        /^modal/,
      ],
    }) : '',
  ],
};
```

I was having a lot of styles missing in production until I added the `'./node_modules/bootstrap-vue/src/components/**/*.js',` line.. as expected

I thought I had solved all issues but just now I ran into a new issue with the Bootstrap Vue Modals. I set it `size: 'xl'` which adds the class `modal-xl` and it worked perfectly on dev, but on production, it was as if that class didn't exist.

Only after I added `/^modal/,` in the whitelistPatterns did I solve the problem. I found it odd that I had to do this as everything else seemed to work.

I don't want to run into any other surprises so if you happen to know what is the correct way to configure Bootstrap Vue with `postcss-purgecss` so that no styles ever go missing, it would be greatly appreciated if you could share it here

Thank you very much : )

**Comment by tmorehouse** [comment]
CSS minification only occurs during production builds (typically), depending on how your project is set up.

purge-css looks for text strings in js and html modules (and Vue SFC files).  it expects the strings to be present in full form, and cannot handle things like `'modal-'+(this.size)` where the classes are constructed at runtime.  This happens for many classes, such as color variants, sizes, etc, as these values cannot be hard coded in the components as they are all configurable via SCSS.

purge-css is great for static projects (html files, etc), but is not the greatest thing to use with things that need to build classes dynamically at run time.

Your best bet is to forgo using purge-css, and just import the source Bootstrap SCSS files that are needed (it is rather modular in structure: https://github.com/twbs/bootstrap/tree/v4-dev/scss), and similar for BootstrapVue's SCSS.

There are also many SASS variables that can control how much CSS is generated.  If you don't need all the theme colors, you can provide a smaller SASS map of the theme colors needed, etc.

**Comment by vesper8** [comment]
Thanks for that @tmorehouse 

The thing is this is a rather big project that also makes use of Tailwind CSS and I've noticed that when using postcss-purgecss my css bundle size is dramatically lower so I would prefer to keep using it and instead build a proper whitelist that will account for all of the dynamically generated classes.. is there something specific I could search the bootstrap vue codebase to identify all of the classes that can be generated dynamically such as `'modal-'+(this.size)` ? Even if that list grew to dozens of entries, it would still be preferable to use it this way then to not use postcss-purgecss at all.

**Comment by tmorehouse** [comment]
Here are some common prefixes (as regular expressions):

Bootstrap Components:
- `/^badge.*/`
- `/^breadcrumb.*/`
- `/^btn.*/`
- `/^btn-group.*/`
- `/^card.*/`
- `/^carousel.*/`
- `/^form-.*/` (default styled form controls)
- `/^custom.*/` (custom styled form controls)
- `/^modal.*/`
- `/^table.*/`
- `/^container.*/`
- `/^row.*/`
- `/^col.*/`
- etc

Bootstrap [Utility classes](https://bootstrap-vue.js.org/docs/reference/utility-classes):
- `/^p.*/` (padding)
- `/^m.*/` (margin)
- `/^w.*/` (width)
- `/^bg-.*/` (background colors)
- `/^border-.*/` (border utilities)
- `/^text-.*/` (text utilities/colors)
- `/^align-.*/` (alignment utilities)
- `/^justify.*/` (justification utilities)
- etc

BootstrapVue Custom components (which may use some of the above classes):
- `/^b-avatar.*/`
- `/^b-sidebar.*/`
- `/^b-form.*/`
- `/^b-table.*/`
- `/^b-toast*/`
- etc

Most of the custom components have an implementation note new the bottom of the page which describes the base classes used, and if custom CSS/SCSS is used.

Most of the sizing classes (btn-sm, form-control-lg, modal-sm, etc) have the end prefixed with either `-sm` or `-lg`. If you are not using these classes, you can blacklist them (just note the date/time picker controls use (btn-sm).

The alternative option is to whitelist the bootstrap SCSS files/directories and BootstrapVue files (note BootstrapVue has SASS variables to disable generation of the custom styles form things like stacked tables, etc).  And then `@import` only the sections of bootstrap SCSS needed (they are organized by component and utilities).

**Comment by vesper8** [comment]
thank you for the boss answer.. as usual : )
